### ISSUE TITLE
client_secret being passed in post data in Basic auth

### ISSUE THREAD

**Issue by carlbarrdahl** [nextauthjs/next-auth#950]
**Describe the bug**
We're trying to integrate Signicat OpenID with Basic auth and get this message:
```
{
  statusCode: 400,
  data: '{"error":"invalid_request","error_description":"Your client is configured to authenticate using CLIENT_SECRET_BASIC"}'
}
```

This is because `client_secret` is sent as post data here, even if its `null` or `undefined`: 
https://github.com/nextauthjs/next-auth/blob/9dbd372f08008463292576fbea8989acef506be3/src/server/lib/oauth/callback.js#L225

With empty client secret (note the: `&client_secret=`):
```js
// formData is: grant_type=authorization_code&client_id=client_id&client_secret=&code=code&redirect_uri=uri
{
  statusCode: 400,
  data: `{"error":"invalid_request","error_description":"Form body malformed! (Body is not a set of key-value pairs separated by '=', delimited by '&' characters)"}`
}
```
Here are some possible solutions:
- Use this querystring library and strip nulls in the query object. https://github.com/sindresorhus/query-string#skipnull
- Change this code to not set the `params.client_secret` at all.

```js
if (!params.client_secret) {
  if (provider.clientSecretCallback) {
    params.client_secret = yield provider.clientSecretCallback(provider.clientSecret);
  } else {
    params.client_secret = provider.clientSecret;
  }
}
```

**Steps to reproduce**
Use custom oauth2 config:
```js
{
  id: "signicat",
  name: "Signicat BankID",
  type: "oauth",
  version: "2.0",
  scope: "openid profile signicat.national_id signicat.certificate",
  params: { grant_type: "authorization_code" },
  accessTokenUrl: `https://${config.signicatHost}/oidc/token`,
  requestTokenUrl: `https://${config.signicatHost}/oidc/token`,
  authorizationUrl: `https://${config.signicatHost}/oidc/authorize?response_type=code&response_mode=form_post&acr_values=urn:signicat:oidc:method:sbid`,
  profileUrl: `https://${config.signicatHost}/oidc/userinfo`,
  clientId: config.signicatClientId,
  clientSecret: config.signicatClientSecret,
  idToken: true,
  state: false,
  clientSecretCallback: () => null, // Remove client_secret from post data in request
  headers: {
    Authorization: `Basic ${Buffer.from(config.signicatClientId + ":" + config.signicatClientSecret, "ascii").toString("base64")}`,
  }
}
```

**Expected behavior**
`client_secret` should be omitted from postData if `headers.Authorization.includes("Basic")`

**Feedback**
*Documentation refers to searching through [online documentation](https://next-auth.js.org), code comments and issue history. The example project refers to [next-auth-example](https://github.com/iaincollins/next-auth-example).*

* [x] Found the documentation helpful
* [ ] Found documentation but was incomplete
* [ ] Could not find relevant documentation
* [x] Found the example project helpful
* [ ] Did not find the example project helpful

**Comment by balazsorban44** [comment]
Hi there!

Just out of curiosity, since I'm not familiar with Signicat, have you actually tried to just provide a secret and see if sending it in post data would work? 

Is this the documentation of what you are referring to?: https://developer.signicat.com/documentation/authentication/protocols/openid-connect/endpoints/

I think we have to see if we want to/need to expand the public facing API as you suggested above, but I have to discuss this further first.

I'll get back to you with more information, 

Thanks!

**Comment by carlbarrdahl** [comment]
Hi @balazsorban44. Thanks for your response.

Yes when sending the `client_secret` Signicat throws me the first error (Your client is configured to authenticate using CLIENT_SECRET_BASIC).

I was able to get around this by monkeypatching querystring to filter out empty keys:
```js
const _stringify = querystring.stringify;
querystring.stringify = (...args) => {
  args[0] = Object.keys(args[0]).reduce(
    (obj, x) => (args[0][x] ? { ...obj, [x]: args[0][x] } : obj),
    {}
  );
  return _stringify.call(null, ...args);
};
```

However, I got stuck again when trying to parse the profileData:
https://github.com/nextauthjs/next-auth/blob/9dbd372f08008463292576fbea8989acef506be3/src/server/lib/oauth/callback.js#L144

Since we're using encryption Signicat is sending back an encrypted JWT which obviously fails to `JSON.parse`.

This is unfortunate because if I could only get past that line, I would be able to decode it in the `provider.profile` function.
https://github.com/nextauthjs/next-auth/blob/8827950f12ffd594a9b60fc23a264266b12530e9/src/server/lib/oauth/callback.js#L153

I really wish to use this excellent library but for now I don't see how to get it working without doing a fork and rewrite parts of:
`/src/server/lib/oauth/callback.js` to change:
- Omit `client_secret` from post data if `headers.Authorization` is set in `_getOAuthAccessToken`
- Allow encrypted profileData (don't throw if `JSON.parse` fails) in `_getProfile`

**Comment by balazsorban44** [comment]
Hmm, if you have the time and interest, I think you could try create the changes and propose it in a PR, and we can discuss further what we could incorporate into the core package. I think if it is general enough solution that we can back up with links to official OAuth and/or OIDC specs and doesn't alter the user facing api much (or at all), I am pretty sure we could accept such a change.

**Comment by carlbarrdahl** [comment]
After playing around with the code base and with [panva/node-openid-client](https://github.com/panva/node-openid-client), I think perhaps this could be a solution:

- Provide a way to add a private keystore to provider config
- Set `provider.idToken` to an object with encryption algorithms 
- Decrypt idToken if keystore and algorithms are present

Here is the relevant code in panva/node-openid
https://github.com/panva/node-openid-client/blob/82382e8e8904e9cfb87228d11fd01ca7c425af9d/lib/client.js#L449-L457

Basically there needs to be a way to decrypt tokens before they get decoded if keys are present.

**Comment by mod-flux** [comment]
@balazsorban44 I've just come across this issue with an OAuth service that doesn't accept via POST. As per the most up-to-date IETF OAuth 2.0 Authorization Framework spec, providing the client credentials via the request-body (POST) is explicitly called out as **NOT RECOMMENDED** and only to be used in situations where clients are unable to directly utilize the HTTP Basic authentication scheme, which doesn't cover this library. Here is the snippet in question in the RFC:

> Including the client credentials in the request-body using the two
   parameters is NOT RECOMMENDED and SHOULD be limited to clients unable
   to directly utilize the HTTP Basic authentication scheme (or other
   password-based HTTP authentication schemes).  The parameters can only
   be transmitted in the request-body and MUST NOT be included in the
   request URI.

Reference: [https://tools.ietf.org/html/rfc6749#section-2.3.1](https://tools.ietf.org/html/rfc6749#section-2.3.1)

Not having the ability to choose between Basic Auth and POST by easily stripping out the `client_id` and `client_secret` from the request body and private it via  Basic Auth is a blocker for us as our OAuth provider is following the latest OAuth specifications and doesn't marry with the support available with NextAuth. :-(

**Comment by Fronix** [comment]
@carlbarrdahl We have the exact same issue using Signicat's OpenID-connect solution. Did you manage to get it working with the modifications?

**Comment by balazsorban44** [comment]
Please have a look at #1846 and tell me if it could help!

**Comment by balazsorban44** [comment]
A hacky way is to set `id: "reddit"` in the Provider config for now as I found out yesterday:

https://github.com/nextauthjs/next-auth/blob/0fae0c7a8eb224964e38da409ecb9ba0497e8c50/src/server/lib/oauth/client.js#L144-L146

#1846 would hopefully help to deal with this more elegantly

**Comment by mod-flux** [comment]
> A hacky way is to set `id: "reddit"` in the Provider config for now as I found out yesterday:
> 
> https://github.com/nextauthjs/next-auth/blob/0fae0c7a8eb224964e38da409ecb9ba0497e8c50/src/server/lib/oauth/client.js#L144-L146
> 
> #1846 would hopefully help to deal with this more elegantly

Yeap @balazsorban44 that gets us halfway there, the only other thing that's needed is removing `client_secret` from `params` before this line:

https://github.com/nextauthjs/next-auth/blob/0fae0c7a8eb224964e38da409ecb9ba0497e8c50/src/server/lib/oauth/client.js#L156

Otherwise, the OAuth vendor complains that you're sending through the client credentials in an invalid format and rejects the request.

**Comment by Fronix** [comment]
> > A hacky way is to set `id: "reddit"` in the Provider config for now as I found out yesterday:
> > https://github.com/nextauthjs/next-auth/blob/0fae0c7a8eb224964e38da409ecb9ba0497e8c50/src/server/lib/oauth/client.js#L144-L146
> > 
> > #1846 would hopefully help to deal with this more elegantly
> 
> Yeap @balazsorban44 that gets us halfway there, the only other thing that's needed is removing `client_secret` from `params` before this line:
> 
> https://github.com/nextauthjs/next-auth/blob/0fae0c7a8eb224964e38da409ecb9ba0497e8c50/src/server/lib/oauth/client.js#L156
> 
> Otherwise, the OAuth vendor complains that you're sending through the client credentials in an invalid format and rejects the request.

There is another way to get you there using [patch-package](https://www.npmjs.com/package/patch-package), modify the `client.js` file and make a patch.

```javascript
  // Added as a fix for Reddit Authentication
  if (provider.id === 'reddit' || provider.id === 'mycustomprovider') {
    headers.Authorization = 'Basic ' + Buffer.from((provider.clientId + ':' + provider.clientSecret)).toString('base64')
  }
  if(provider.id === 'mycustomprovider') {
    delete params.client_secret;
  }
```

This is how we solved it for now.

**Comment by balazsorban44** [comment]
I Will close this in favor of #1846

**Comment by wowczarczyk** [comment]
In case anyone wanders in here trying to use Signicat Express with `next-auth`, here is a config that worked for me:

```{
      id: "signicat",
      name: "Signicat BankID",
      type: "oauth",
      version: "2.0",
      wellKnown:
        "https://login-test.signicat.io/.well-known/openid-configuration",
      token: {
        params: {
          scope: "openid nin phone email profile",
        },
      },
      client: {
        redirect_uris: ["http://localhost:3000/api/auth/callback/signicat"],
        client_id: "**********",
        client_secret:
          "************",
      },
      checks: ["pkce", "nonce"],

      issuer: "https://login-test.idfy.no",

      clientId: "**********",
      clientSecret:
        "***********",
      idToken: true,
      profile: (profile) => {
        return {
          id: profile.sub,
          name: profile.name,
          email: profile.email,
          nin: profile.nin,
          phone: profile.phone,
        };
      },
      jwks_endpoint:
        "https://login-test.signicat.io/.well-known/openid-configuration/jwks",
      authorization: {
        url: "https://login-test.signicat.io/connect/authorize",
        params: {
          scope: "openid nin phone email profile",
          grant_type: "authorization_code",
          redirect_uri: "http://localhost:3000/api/auth/callback/signicat",
        },
      },
    },
    ```
